Basic Chromium backend implementation

duswnchl opened the pull request:
This is a basic implementation of Chromium backend and Wolvic communicates with Chromium directly via Content API. Also, this is working in progress and you need prebuilt AARs to run Wolvic with Chromium backend.

duswnchl commented on app/build.gradle:
```diff
@@ -517,6 +532,13 @@ android {
+            noCompress 'dat'
+            noCompress 'bin'
+            noCompress 'pak'
+        }
```
note: I cannot find a good solution to fully explicit these options from other variants. Does anybody has idea?

duswnchl replied:
`userProperties.backend` is deprecated #491. Sadly, flavor dimension `backend` doesn't work for aaptOptions as well. I tried to `onVariant` inside `androidComponents` block but it is also failed. I guess that there would be better solution after  upgrading `android_gradle_plugin`.

zakharvoit replied:
OK, Let's add a TODO here if possible. I'm fine with the way it is, since we don't have any non-chromium assets that would be affected.

duswnchl commented on app/src/common/shared/com/igalia/wolvic/browser/api/WFactory.java:
```diff
@@ -19,7 +19,12 @@ public class WFactory {
+
+    public static @NonNull WSession createSession(
+            @Nullable WSessionSettings settings, @NonNull Context context) {
+        return new SessionImpl(settings, context);
```
note: The overload of createSession is added because current chromium implementation needs Activity context to initialize window.

duswnchl replied:
Removed.

svillar commented on app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java:
```diff
@@ -207,27 +151,21 @@ public String getDefaultUserAgent(int mode) {
     public void releaseDisplay(@NonNull WDisplay display) {
-        assert mDisplay != null;
-        mRuntime.releaseDisplay(mDisplay.getBrowserDisplay());
-        mDisplay = null;
```
Maybe a `TODO` is required here? Or does it not make much sense? In any case we need a clarification.

duswnchl replied:
I added `TODO` comments here.

svillar commented on app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TextInputImpl.java:
```diff
@@ -13,7 +13,7 @@
 
 public class TextInputImpl implements WTextInput {
-    SessionImpl mSession;
+    WSession mSession;
```
Why this? I mean, at this level we can refer to the SessionImpl if we want, just curious

duswnchl replied:
Oops. This is my mistake and redundant fix. Removed in the latest PR.

svillar commented on app/src/common/chromium/org/chromium/components/embedder_support/view/WolvicContentRenderView.java:
```diff
@@ -0,0 +1,48 @@
+                        PixelFormat.OPAQUE,
+                        width,
+                        height,
+                        surface);
```
Ditto about using a much more compact representation

duswnchl replied:
Done.

svillar reviewed (changes requested):
Looking great. Most of my comments should be easy to fix. The main blocker at this point is adding the CHROMIUM BuildOption which should not be needed.

Ah, and also this should not break the current Gecko builds.

zakharvoit commented on app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java:
```diff
@@ -9,42 +11,53 @@
+                if (delegate != null) {
+                    delegate.onFirstComposite(mSession);
+                }
+            }, 100);
```
I believe we don't need this anymore. This has been addressed in the following PR: https://github.com/Igalia/wolvic/pull/514. I'm not sure if we want to address it here though, because we will be rebasing these in-progress CLs on top when this is merged.

duswnchl replied:
IMO, this statement should be cleaned here since it is for the weblayer implementation. Removed.

duswnchl commented:
Thanks for comments and I can update this PR that addressed most comments. Let me set it draft while working for it.

duswnchl commented on app/src/common/shared/com/igalia/wolvic/browser/engine/Session.java:
```diff
@@ -563,8 +564,10 @@ private WSession createWSession(@NonNull SessionSettings aSettings) {
-        WSession session = WFactory.createSession(settings);
-        setupSessionListeners(session);
+        WSession session =
+                BuildConfig.CHROMIUM
```
I understand your concern and this part is removed in the latest PR.

duswnchl commented on app/build.gradle:
```diff
@@ -122,6 +122,7 @@ android {
         buildConfigField "String[]", "SPEECH_SERVICES", "{ com.igalia.wolvic.speech.SpeechServices.MEETKAI }"
         buildConfigField "Boolean", "SUPPORTS_SYSTEM_NOTIFICATIONS", "false"
         buildConfigField "Boolean", "KIOSK_MODE_ALWAYS", "false"
+        buildConfigField "Boolean", "CHROMIUM", "false"
```
Removed. Done.

duswnchl commented on app/build.gradle:
```diff
@@ -373,9 +375,13 @@ android {
+            if (!platform.startsWith('oculusvr'))
+                variant.setIgnore(true);
+            else
+                variant.setIgnore(backend != 'gecko')
```
Aha! I thought that this change would be need because we didn't test chromium backend with store variants. Thanks for sharing context and this change is removed.

duswnchl commented on app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java:
```diff
@@ -9,42 +11,53 @@
     public void surfaceChanged(@NonNull Surface surface, int width, int height) {
         // Make sure we update the current active tab. We have an empty SessionImpl.setActive implementation
         // because in weblayer the browser window is selected from a pool. We can use the hint that
         // the session want's to be rendered into this surface to set it active too.
```
Makes sense. Removed.

duswnchl commented on app/src/common/chromium/com/igalia/wolvic/browser/api/impl/RuntimeImpl.java:
```diff
@@ -1,108 +1,111 @@
+    }
+
+    private void init(Application application) {
+        boolean isBrowserProcess = !ContextUtils.getProcessName().contains(":");
```
I added more details in the method name to work. (i.e. initBrowserProcess)

duswnchl commented on app/src/common/chromium/com/igalia/wolvic/browser/api/impl/RuntimeImpl.java:
```diff
@@ -1,108 +1,111 @@
-        BrowserDisplay display = new BrowserDisplay(mContext);
-        display.attach(mFragmentManager, mViewContainer, fragment);
-        return display;
+                            @Override
```
Done.

duswnchl commented on app/src/common/chromium/com/igalia/wolvic/browser/api/impl/RuntimeImpl.java:
```diff
@@ -1,108 +1,111 @@
+                            @Override
+                            public void onSuccess() {
+                                Log.i(LOGTAG, "Browser started!");
+                            }
```
Got it. Done.

duswnchl commented on app/src/common/chromium/com/igalia/wolvic/browser/api/impl/RuntimeImpl.java:
```diff
@@ -1,108 +1,111 @@
     public WResult<Void> clearData(long flags) {
-        // TODO: Implement
-        return WResult.fromValue(null);
+        return null;
```
It's better to keep original codes. Done.

duswnchl commented on app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java:
```diff
@@ -207,27 +151,21 @@ public String getDefaultUserAgent(int mode) {
-        mDisplay = new DisplayImpl(mRuntime.acquireDisplay(mSettings.getUsePrivateMode()), this);
-        mDisplay.getBrowserDisplay().getBrowser().addTab(mTab);
-        mDisplay.getBrowserDisplay().getBrowser().setActiveTab(mTab);
+        WebContents webContents = mRuntime.createWebContents();
```
Makes sense. Done.

duswnchl reviewed (commented):
Hi! This PR is separated in 2 commits. PTAL.
- Remove WebLayer backend implementation
- Basic Chromium backend implementation

duswnchl commented on app/src/common/chromium/com/igalia/wolvic/browser/api/impl/RuntimeImpl.java:
```diff
@@ -35,12 +54,53 @@ public RuntimeImpl(@NonNull Context ctx, @NonNull WRuntimeSettings settings) {
+
+        ContextUtils.initApplicationContext(context);
+        ResourceBundle.setNoAvailableLocalePaks();
+        // Native libraries for child processes are loaded in its implementations in content.
```
note: FYI -> https://source.chromium.org/chromium/chromium/src/+/refs/heads/main:content/public/android/java/src/org/chromium/content/app/ContentChildProcessServiceDelegate.java;drc=2450f2f5d0ce0da9b8cf493c533f9528ff17bab6;l=61

svillar reviewed (changes requested):
I added a couple of comments to the code. I also have a more general doubt, in the previous version of the patch there was some code added to WFactory to modify how the SessionImpl was created for the Chromium flavour, I don't see any change related to that in this PR, is it because it's going to be resolved in a future PR?

duswnchl commented:
> I added a couple of comments to the code. I also have a more general doubt, in the previous version of the patch there was some code added to WFactory to modify how the SessionImpl was created for the Chromium flavour, I don't see any change related to that in this PR, is it because it's going to be resolved in a future PR?

That codes are no longer needed and removed in this PR because we can monitor activity status by setting application context at chromium layer via chromium API.  This PR doesn't touch any codes in shared/.

duswnchl commented on app/src/common/chromium/com/igalia/wolvic/browser/api/impl/PanZoomControllerImpl.java:
```diff
@@ -0,0 +1,25 @@
+    public void onMotionEvent(@NonNull MotionEvent event) {
+        // TODO: implement
+    }
+}
```
@svillar I just notice that this change is included in https://github.com/Igalia/wolvic/commit/2e37562c6aa247f9e63ea64b142e2c01d27ddf4b. This change is removed in the latest PR.

duswnchl reviewed (commented):
> I added a couple of comments to the code. I also have a more general doubt, in the previous version of the patch there was some code added to WFactory to modify how the SessionImpl was created for the Chromium flavour, I don't see any change related to that in this PR, is it because> I added a couple of comments to the code. I also have a more general doubt, in the previous version of the patch there was some code added to WFactory to modify how the SessionImpl was created for the Chromium flavour, I don't see any change related to that in this PR, is it because it's going to be resolved in a future PR?

 @svillar : Codes in the previous version which was for additional SessionImpl overload for Chromium flavor is no longer needed and has been removed on the latest PR. That overload was for passing the activity context to SessionImpl, but it can cover in the application context of RuntimeImpl.  We don't have any platform specific changes in shared codes now.

svillar commented on app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java:
```diff
@@ -134,6 +162,10 @@ public void exitFullScreen() {
     public WDisplay acquireDisplay() {
         assert mDisplay == null;
         // TODO: implement correctly
```
Remove this comment now

svillar reviewed (changes requested):
Almost there

duswnchl commented on app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java:
```diff
@@ -9,21 +9,35 @@
+
+        try {
+            // Dispatch onSurfaceCreated
+            mRenderView.surfaceCreated(surface);
```
ContentViewRenderView.surfaceCreated should be called when surface is created, but it seems that there isn't explicit callback for surface creation in wolvic. I think that calling mRenderView.surfaceCreated only when mSurface is null would be work here. Let me update PR for it.

svillar reviewed (approved):
Good job!
